### PR TITLE
Add charset and template variables for db creation

### PR DESCRIPTION
**Pull request by stromvirvel**
These changes will add three additional environment variables, to set collation order (LC_COLLATE), character classification (LC_CTYPE) and a template (TEMPLATE). If unset, the container should behave like before (using default values).

See:
https://www.postgresql.org/docs/9.3/static/sql-createdatabase.html

Successfully tested with postgresql:9.3

### CODE REVIEW

File: etc/service/postgresql/run [hunk 1]
```diff
@@ -30,7 +30,7 @@ if ! [[ $(ls -A "${DATADIR}" | grep -v placeholder) ]]; then
     pgsql_command "CREATE ROLE ${PGSQL_ROLE_1_USERNAME} WITH ${PGSQL_ROLE_1_FLAGS} ENCRYPTED PASSWORD '${PGSQL_ROLE_1_PASSWORD}';"
   fi
   if [[ -n "${PGSQL_DB_1_NAME}" && -n "${PGSQL_DB_1_OWNER}" ]]; then
-    pgsql_command "CREATE DATABASE ${PGSQL_DB_1_NAME} WITH OWNER ${PGSQL_DB_1_OWNER} ENCODING '${PGSQL_DB_1_ENCODING:-UNICODE}';"
+    pgsql_command "CREATE DATABASE ${PGSQL_DB_1_NAME} WITH OWNER ${PGSQL_DB_1_OWNER} ENCODING '${PGSQL_DB_1_ENCODING:-UNICODE}' LC_COLLATE '${PSQL_DB_1_LC_COLLATE}' LC_CTYPE '${PSQL_DB_1_LC_CTYPE}' TEMPLATE ${PSQL_DB_1_TEMPLATE:-DEFAULT};"
```
Comment (mitar) [inline etc/service/postgresql/run]: Should you quote template name as well?

**Comment by mitar** [comment]
Are you sure that specifying `LC_COLLATE ''` is the same as not specifying that argument at all?

File: etc/service/postgresql/run (hunk 1 above)
Comment (stromvirvel) [reply etc/service/postgresql/run]: I thought the same, but the examples in these two documents shows unquoted:

- https://www.postgresql.org/docs/9.3/static/sql-createdatabase.html
- https://confluence.atlassian.com/adminjiraserver071/connecting-jira-applications-to-postgresql-802592176.html

**Comment by stromvirvel** [comment]
I'm not 100% sure, but in the postgresql manual it reads that:

> The default is to use the collation order of the template database. See below for additional restrictions.

or:
> The default is to use the character classification of the template database. See below for additional restrictions.

I double checked it by creating a container with your current version and my patched version. Both values were set to `en_US.UTF-8`.

We could make sure and insert an if condition?

**Comment by mitar** [comment]
OK, it seems empty string is default:

https://www.postgresql.org/docs/9.1/static/runtime-config-client.html

And reads from the environment.

There is some discussion about this [here](https://www.postgresql.org/message-id/1331490004.6425.20.camel@jdavis).

Could you please do the following checks:
* create a database without specifying `LC_COLLATE` and `LC_CTYPE` and see what is the the resulting collate and ctype for the database (it should be `en_US.UTF-8`)
* change your local environment variables and set those environment variables to something else, try again to create the database without specifying `LC_COLLATE` and `LC_CTYPE`
* with changed local environment variables, but specifying `LC_COLLATE` and `LC_CTYPE` to empty string

**Comment by stromvirvel** [comment]
**Edit**
I had to re-run all tests with `PGSQL_DB_1_TEMPLATE=template0` set, otherwise postgresql server won't allow to set collate to 'C', with the following error message:
```
ERROR:  new collation (C) is incompatible with the collation of the template database (en_US.UTF-8)
HINT:  Use the same collation as in the template database, or use template0 as template.
```

---

I'm a bit confused about what environment variables you are talking about now. I just interpret what makes sense to me, please double check.

> create a database without specifying LC_COLLATE and LC_CTYPE and see what is the the resulting collate and ctype for the database (it should be en_US.UTF-8)

`docker run --name postgresql -d -p 5432:5432 -e PGSQL_ROLE_1_USERNAME=testuser -e PGSQL_ROLE_1_PASSWORD=testpassword -e PGSQL_DB_1_NAME=testdbname -e PGSQL_DB_1_OWNER=testuser -e PGSQL_DB_1_TEMPLATE=template0 docker-postgresql`

**Result:** Collation and character type set to `en_US.UTF-8`


---

> change your local environment variables and set those environment variables to something else, try again to create the database without specifying LC_COLLATE and LC_CTYPE


`docker run --name postgresql -d -p 5432:5432 -e PGSQL_ROLE_1_USERNAME=testuser -e PGSQL_ROLE_1_PASSWORD=testpassword -e PGSQL_DB_1_NAME=testdbname -e PGSQL_DB_1_OWNER=testuser -e PGSQL_DB_1_TEMPLATE=template0 -e LC_COLLATE=C -e LC_CTYPE=C docker-postgresql`

**Result:** Collation and character type set to `en_US.UTF-8`

---


> with changed local environment variables, but specifying LC_COLLATE and LC_CTYPE to empty string

This is exactly the same as your request before, only with *explicit* passing empty variables ...?


`docker run --name postgresql -d -p 5432:5432 -e PGSQL_ROLE_1_USERNAME=testuser -e PGSQL_ROLE_1_PASSWORD=testpassword -e PGSQL_DB_1_NAME=testdbname -e PGSQL_DB_1_OWNER=testuser -e PGSQL_DB_1_TEMPLATE=template0 -e LC_COLLATE=C -e LC_CTYPE=C -e PGSQL_DB_1_LC_COLLATE= -e PGSQL_DB_1_LC_CTYPE= docker-postgresql`

**Result:** Collation and character type set to `en_US.UTF-8`

---


I did a further test:
`docker run --name postgresql -d -p 5432:5432 -e PGSQL_ROLE_1_USERNAME=testuser -e PGSQL_ROLE_1_PASSWORD=testpassword -e PGSQL_DB_1_NAME=testdbname -e PGSQL_DB_1_OWNER=testuser -e PGSQL_DB_1_TEMPLATE=template0 -e LC_ALL=C -e LANG=de_CH -e LANGUAGE=de_CH docker-postgresql`

**Result:** Collation and character type set to `en_US.UTF-8`

---


Doesn't seem to read from locale settings at all.

These tests were done by a local image called "docker-postgresql" which was build with the patched code. If you need other tests, please clarify. :-)

**Comment by stromvirvel** [comment]
Oops, I made a typo in the variable's name of file "run". I'll fix that and re-run my tests.

**Comment by stromvirvel** [comment]
I re-run the tests and made changes in my original post. Please review.

**Comment by mitar** [comment]
Cool.

**Comment by stromvirvel** [comment]
Nice, thx :)
Off topic: I'm wondering how you set up automated build on hub.docker.com with multiple dockerfiles. Do you have more information about that?

**Comment by mitar** [comment]
You just specify files under Dockerfile location. So you can put path to a file, not just directory.

screen shot 2017-04-27 at 12 33 29 am

**Comment by stromvirvel** [comment]
Ah very cool thank you! Just noticed that you specified 9.3 as latest. Is this intended?

**Comment by mitar** [comment]
This is for backwards compatibility. I do not think anyone should be using image without specifying the tag. But we had `latest` tag be 9.3 when I started with this image, not thinking about the future, so I am leaving it like that. Personally, I think we should just remove it. But then it might break things for people.

So after some time I will probably remove it. Hopefully in meantime people will migrate away from not using a tag.

**Comment by mitar** [comment]
I do not think you want database to upgrade to a new version under you automatically. :-)

**Comment by stromvirvel** [comment]
No of course not. But there are also developing or testing purposes where you just pull the image without specifying a tag, means it will fetch `latest` and the dev/ops might be confused why 9.3 was fetched :-) That's why I asked. For myself, I don't have a use case for the latest tag with postgresql.

**Comment by mitar** [comment]
Yes. But then people keep using it and then I change it and it breaks. :-) So I am not sure what is the best. Leaving it there so that "it just works" or making it clear that you have to specify version.

**Comment by stromvirvel** [comment]
Well all the other repo maintainers do the same: They just point `latest` to the latest stable release of the packaged software. Take a look at mysql:
https://hub.docker.com/_/mysql/

If you'd point `latest` to the very latest version 9.5 in January 2016, you would have updated it to 9.6 when it was released in September 2016. Users should simply know that the `latest` tag can quickly change.

(Everything IMHO).

**Pull request merged**